core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -160,23 +160,80 @@ class OptimizeExecutor(
 
   private val isMultiDimClustering = zOrderByColumns.nonEmpty
 
-  def optimize(): Seq[Row] = {
+  def optimize(isAutoCompact: Boolean = false, targetFiles: Seq[AddFile] = Nil): Seq[Row] = {
     recordDeltaOperation(txn.deltaLog, "delta.optimize") {
       val minFileSize = sparkSession.sessionState.conf.getConf(
         DeltaSQLConf.DELTA_OPTIMIZE_MIN_FILE_SIZE)
-      val maxFileSize = sparkSession.sessionState.conf.getConf(
-        DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
+      val maxFileSize = if (isAutoCompact) {
+        sparkSession.sessionState.conf.getConf(DeltaSQLConf.AUTO_COMPACT_MAX_FILE_SIZE)
+      } else {
+        sparkSession.sessionState.conf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
+      }
       require(minFileSize > 0, "minFileSize must be > 0")
       require(maxFileSize > 0, "maxFileSize must be > 0")
 
-      val candidateFiles = txn.filterFiles(partitionPredicate)
+      val minNumFilesInDir = if (isAutoCompact) {
+        val minNumFiles =
+          sparkSession.sessionState.conf.getConf(DeltaSQLConf.AUTO_COMPACT_MIN_NUM_FILES)
+        require(minNumFiles > 0, "minNumFiles must be > 0")
+        minNumFiles
+      } else if (isMultiDimClustering) {
+        1
+      } else {
+        // compaction
+        2
+      }
+
+      if (txn.readVersion == -1) {
+        throw DeltaErrors.notADeltaTableException(txn.deltaLog.dataPath.toString)
+      }
+
+      val candidateFiles = if (!isAutoCompact) {
+        txn.filterFiles(partitionPredicate)
+      } else {
+        val autoCompactTarget =
+          sparkSession.sessionState.conf.getConf(DeltaSQLConf.AUTO_COMPACT_TARGET)
+        // Filter the candidate files according to autoCompact.target config.
+        autoCompactTarget match {
+          case "table" =>
+            txn.filterFiles()
+          case "commit" =>
+            targetFiles
+          case "partition" =>
+            val eligiblePartitions = targetFiles.map(_.partitionValues).toSet
+            txn.filterFiles().filter(f => eligiblePartitions.contains(f.partitionValues))
+          case _ =>
+            logError(s"Invalid config for autoCompact.target: $autoCompactTarget. " +
+              s"Use the default value 'table'.")
+            txn.filterFiles()
+        }
+      }
+
       val partitionSchema = txn.metadata.partitionSchema
 
       // select all files in case of multi-dimensional clustering
       val filesToProcess = candidateFiles.filter(_.size < minFileSize || isMultiDimClustering)
-      val partitionsToCompact = filesToProcess.groupBy(_.partitionValues).toSeq
-
-      val jobs = groupFilesIntoBins(partitionsToCompact, maxFileSize)
+      val partitionsToCompact = filesToProcess
+        .groupBy(_.partitionValues)
+        .filter(_._2.size >= minNumFilesInDir)
```

Comment:
Avoid `_._2` as it says pretty much nothing (and may confuse newborn Scala devs). It'd be helpful to have proper names like `filter { case (_, abc) => ... }`